### DISCUSSION TITLE
Why axios doesnt send cookie even if i inset {withCredentails: true} (GetServerSideProps)

### DISCUSSION THREAD

**Discussion by Choco-milk-for-u** [vercel/next.js#37335] [Help] [answered]
i don`t know why, but without it (headers - line), in the server-side it  just return Object: null but with it just {}
```
export const getServerSideProps: GetServerSideProps = async (context) => {
 const { data } = await axios.get('http://localhost:5000/api/authorization/refresh', { withCredentials: true, headers: { cookie: `token ${context.req.cookies['token']}` } });
  return {
    props: {}
  }
}
```

**Comment by icyJoseph** [comment, marked as answer]
Hi,

You shouldn't reach out to API routes from GSSP. It'll be hard to maintain, and open itself to bugs.

Think of it this way. GetServerSideProps, runs already on the server. Why would you relay to yet another server side call, when you could just reach out to the data source directly from GetServerSideProps.

Also, why would axios carry browser cookies inside getServerSideProps if, again, it runs on the server, without any knowledge of the browser.

The correct pattern here is to first, create reusable functions that fetch/axios, from the data source, and use those in `api/routes` or GetServerSideProps, and perhaps even GetStaticProps. 

In short, whatever logic is happening in `api/authorization/refresh`, needs to be abstracted away, so that it can be both used in GetServerSideProps and `api/authorization/refresh`.

#### Example

```ts
// lib/codewars.ts
export const getCodeWarsUser = (user) =>
  axios
    .get(`https://www.codewars.com/api/v1/users/${user}`)
    .then(({ data }) => data);
```

Then in getServerSideProps

```ts
import {getCodeWarsUser} from 'lib/codewars';

export async function getServerSideProps() {
  const codewars = await getCodeWarsUser('some-user');

  return {
    props: { codewars },
    revalidate: 10,
  };
}
```

And also in api/codewars

```ts
import {getCodeWarsUser} from 'lib/codewars';

const codewars = async (req, res) => {
  const data = await getCodeWarsUser('some-user');

  return res.send(data);
};

export default codewars;
```

If you are worried about leaking data to the client by writing Server Side code in GetServerSideProps (which is part of a page), then you can use this tool https://next-code-elimination.vercel.app/ to help you see, what is actually sent to the client, as a JavaScript bundle.

**Comment by Choco-milk-for-u** [comment]
Hello thank you for your answer it was helpful. I just thought i can use getServerSideProps for before rendering page, get acces jwt-token. And every time i refresh this page i wanted get value that saying user is authorizated. Now after your answer i guess better choice is to use useEffect. and make pre-loader component

**Reply by icyJoseph** [reply]
You can still do that from gSSP, just connect to the service directly. Just take the cookie from the incoming request.
